### PR TITLE
Feature/ablan abkenov/crud operations frontend

### PR DESCRIPTION
**Pull request by abkenov**
Created components for CRUD operations for Todos, and some authorization components
Forms created with Formik, Validation conducted with Yup, Fetching done by React Query, Requests are in a separate class

Some issues with TS, could not assign type for destructured props
interface design is to be improved

### CODE REVIEW

File: mobile/App.tsx [hunk 1]
```diff
@@ -1,21 +1,36 @@
 import { StatusBar } from 'expo-status-bar';
 import React from 'react';
```
Comment (CodeGeneration-2020) [inline mobile/App.tsx:2]: this import should go first

File: mobile/App.tsx [hunk 2]
```diff
@@ -1,21 +1,36 @@
...
-    <View style={styles.container}>
-      <Text>We are Live!</Text>
-      <StatusBar style="auto" />
-    </View>
+    <NavigationContainer>
+      <QueryClientProvider client={queryClient}>
+        <SafeAreaView style={styles.container}>
+          <StatusBar style="auto" />
+          <Tabs />
+          {/* <Authorization /> */}
```
Comment (CodeGeneration-2020) [inline mobile/App.tsx:22]: remove commented code

File: mobile/App.tsx [hunk 3]
```diff
@@ -1,21 +1,36 @@
...
   );
 }
 
 const styles = StyleSheet.create({
   container: {
     flex: 1,
     backgroundColor: '#fff',
-    alignItems: 'center',
-    justifyContent: 'center',
+    // alignItems: 'center',
```
Comment (CodeGeneration-2020) [inline mobile/App.tsx:33]: remove commented code

File: mobile/components/EditTodo.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import TodoService from '../services/todo.service';
+import * as yup from 'yup';
+
+const todo_service = new TodoService();
+
+export default function EditTodo({ navigation, route }) {
+
+  const { id } = route.params
+
+  const validationSchema = yup.object().shape({
```
Comment (CodeGeneration-2020) [inline mobile/components/EditTodo.tsx:14]: please move validation schemas to a separate file

File: mobile/components/Registration.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    username: yup.string().typeError('string').required('Username field can not be empty!'),
+    email: yup.string().email('Appropriate e-mail is required!').required('E-mail field can not be empty'),
+    password: yup.string().typeError('string').required('Password field can not be empty!'),
+    passwordVerification: yup.string().oneOf([yup.ref('password')], 'Passwords do not match!').required('Password verification is needed!'),
+  })
+
+  return (
+    <SafeAreaView style={styles.container}>
+      <Formik
+        initialValues={
```
Comment (CodeGeneration-2020) [inline mobile/components/Registration.tsx:18]: can we move it to a separate and `REUSABLE` object? theoretically

File: mobile/components/Registration.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                    onChangeText={handleChange('passwordVerification')}
+                    onBlur={handleBlur('passwordVerification')}
+                    value={values.passwordVerification}
+                    secureTextEntry={true}
+                />
+                {touched.passwordVerification && errors.passwordVerification && <Text>{errors.passwordVerification}</Text>}
+              </ScrollView>
+            </TouchableWithoutFeedback>
+          <Button 
+            onPress={() => {
```
Comment (CodeGeneration-2020) [inline mobile/components/Registration.tsx:76]: please avoid inline functions. everything should be named

File: mobile/components/SignIn.tsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  return (
+    <SafeAreaView style={styles.container}>
+      <Formik
+        initialValues={
+        { 
+          username: '',
+          password: '',
+        }
+      }
+        onSubmit={values => console.log(values)}
```
Comment (CodeGeneration-2020) [inline mobile/components/SignIn.tsx:16]: what is the purpose of `console.log` here? remove it

File: mobile/components/SignIn.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
+import React from 'react';
+import { Button, SafeAreaView, StyleSheet, Text, TextInput, ScrollView, KeyboardAvoidingView, TouchableWithoutFeedback, Keyboard, Platform } from 'react-native';
+import { Formik } from 'formik';
+
+export default function SignIn({ navigation }) {
+
+  return (
+    <SafeAreaView style={styles.container}>
+      <Formik
+        initialValues={
```
Comment (CodeGeneration-2020) [inline mobile/components/SignIn.tsx:10]: can we move it to a separate and `reusable` object?

File: mobile/components/SignIn.tsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
+import React from 'react';
+import { Button, SafeAreaView, StyleSheet, Text, TextInput, ScrollView, KeyboardAvoidingView, TouchableWithoutFeedback, Keyboard, Platform } from 'react-native';
```
Comment (CodeGeneration-2020) [inline mobile/components/SignIn.tsx:2]: this string is too long. please distribute imports into several lines

File: mobile/components/SignIn.tsx [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                    style={styles.textInput}
+                    onChangeText={handleChange('password')}
+                    onBlur={handleBlur('password')}
+                    value={values.password}
+                    secureTextEntry={true}
+                />
+              </ScrollView>
+            </TouchableWithoutFeedback>
+          <Button 
+            onPress={() => handleSubmit()}
```
Comment (CodeGeneration-2020) [inline mobile/components/SignIn.tsx:43]: please avoid inline functions. everything should be named

File: mobile/components/SignIn.tsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+                />
+              </ScrollView>
+            </TouchableWithoutFeedback>
+          <Button 
+            onPress={() => handleSubmit()} 
+            title="Sign In"
+            disabled={!isValid || !dirty}
+          />
+          <Button 
+            onPress={() => navigation.push('Registration')}
```
Comment (CodeGeneration-2020) [inline mobile/components/SignIn.tsx:48]: please avoid inline functions. everything should be named

File: mobile/components/Tabs.tsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+const Tab = createBottomTabNavigator();
+const NewTodoStack = createStackNavigator();
+const TodoContainerStack = createStackNavigator();
+
+const TodoContainerStackScreen = ({ navigation }) => (
+  <TodoContainerStack.Navigator>
+    <TodoContainerStack.Screen name='TodoContainer' component={TodoContainer} options={{
+          headerRight: () => (
+            <Button
+              onPress={() => navigation.navigate('CreateTodoScreen')}
```
Comment (CodeGeneration-2020) [inline mobile/components/Tabs.tsx:19]: please avoid inline functions. everything should be named

File: mobile/components/Tabs.tsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+const TodoStackScreen = () => (
+  <NewTodoStack.Navigator>
+    <NewTodoStack.Screen name='CreateTodoScreen' component={CreateTodo} />
+  </NewTodoStack.Navigator>
+)
+
+export default function Tabs() {
+  return (
+      <Tab.Navigator>
+        <Tab.Screen name="Todos" component={TodoContainerStackScreen} options={{headerShown: false}} />
```
Comment (CodeGeneration-2020) [inline mobile/components/Tabs.tsx:39]: spacings between curly braces

File: mobile/components/TodoContainer.tsx [hunk 14]
```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import { SafeAreaView, StyleSheet, Text, FlatList, Button } from 'react-native';
+import { useQuery } from 'react-query';
+import TodoService from '../services/todo.service';
+
+const todo_service = new TodoService();
+
+const TodoElement = ({title, description, id, date, isCompleted, isPublic, navigation}) => {
```
Comment (CodeGeneration-2020) [inline mobile/components/TodoContainer.tsx:8]: spacings between curly braces

File: mobile/components/TodoContainer.tsx [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  return(
+    <>
+      <Text>title: {title}</Text>
+      <Text>description: {description}</Text>
+      <Text>date: {date}</Text>
+      <Text>isCompleted: {`${isCompleted}`}</Text>
+      <Text>isPublic: {`${isPublic}`}</Text>
+      <Button title={'edit'} onPress={() => navigation.push('EditTodoScreen', {id: id})}></Button>
```
Comment (CodeGeneration-2020) [inline mobile/components/TodoContainer.tsx:16]: please avoid inline functions. everything should be named

File: mobile/components/TodoContainer.tsx [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+      <Button title={'delete'} onPress={() => todo_service.deleteTodoById(id)}></Button>
```
Comment (CodeGeneration-2020) [inline mobile/components/TodoContainer.tsx:17]: please avoid inline functions. everything should be named

File: mobile/components/TodoContainer.tsx [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    </>
+  )
+}
+
+export default function TodoContainer({ navigation }) {
```
Comment (CodeGeneration-2020) [inline mobile/components/TodoContainer.tsx:22]: I would avoid having 2 different components into 1 file?
please always try to separate your business logic into some consolidated modules

File: mobile/components/TodoContainer.tsx [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+  const { data, status } = useQuery('todos', async () => {
+    const { data } = await todo_service.getTodos();
+    return data;
+  });
+  console.log(data)
```
Comment (CodeGeneration-2020) [inline mobile/components/TodoContainer.tsx:28]: remove it

File: mobile/components/TodoContainer.tsx [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+export default function TodoContainer({ navigation }) {
+
+  const { data, status } = useQuery('todos', async () => {
+    const { data } = await todo_service.getTodos();
+    return data;
+  });
+  console.log(data) 
+
+  const renderItem = ( {item} : {item: any} ) => {
```
Comment (CodeGeneration-2020) [inline mobile/components/TodoContainer.tsx:30]: spacings between curly braces are missed

File: mobile/components/TodoElement.tsx [hunk 20]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import { SafeAreaView, StyleSheet, Text, FlatList, View } from 'react-native';
+
+export default function TodoElement(title: String) {
+  return(
+    <>
```
Comment (CodeGeneration-2020) [inline mobile/components/TodoElement.tsx:6]: we can avoid having `<>` here at all, I believe

File: server/src/routes/api/todos.route.ts [hunk 21]
```diff
@@ -1,9 +1,24 @@
...
+    return [
+      check("title", "Title should not be empty").isLength({ min: 1 }),
+      check("content", "Content should not be empty").isLength({ min: 1 })
+    ];
+  }
+}
+
+const todoValidation = new Validation;
+
+
```
Comment (CodeGeneration-2020) [inline server/src/routes/api/todos.route.ts:18]: redundant line

**Review by CodeGeneration-2020 — CHANGES_REQUESTED** [review]
- Please configure `ESlint` for all the files all over the app
- Always avoid inline functions. Everything should be named, always
- Do not heep 2 components, logical functions into 1 file. Separate as much as you can in order to increase readability

File: mobile/components/CreateTodo.tsx [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+
+          <Text>isComplete</Text>
+            <BouncyCheckbox 
+              onPress={() => values.isCompleted = !values.isCompleted}>
+            </BouncyCheckbox>
+
+          <Button 
+            onPress={() => handleSubmit()} 
+            title="Submit"
+            disabled={!isValid || !dirty}
```
[the remark this answers is not in the document]
Comment (abkenov) [reply mobile/components/CreateTodo.tsx:77]: It's a defined prop from Formik API, it checks if our form has been touched or not

File: mobile/components/CreateTodo.tsx [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  const toggleIsCompleted = (values: any) => {
+    values.isCompleted = !values.isCompleted
+  }
+
+  return (
+    <Formik
+      initialValues={initialValues}
+      onSubmit={values => {
```
Comment (CodeGeneration-2020) [inline mobile/components/CreateTodo.tsx:31]:
```
const onFormSubmit = () => {
   toDoService.createNewTodo(values)
   navigation.navigate('TodoContainer')
}
```

&& 

```
onSubmit={onFormSubmit}
```

Comment (CodeGeneration-2020) [reply mobile/components/CreateTodo.tsx:31]: Everything should be named!

File: mobile/components/CreateTodo.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (CodeGeneration-2020) [reply mobile/components/CreateTodo.tsx:78]: @abkenov leave comments in order to explain it

File: mobile/components/CreateTodo.tsx [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+          <TextInput
+            style={styles.input}
+            onChangeText={handleChange('year')}
+            onBlur={handleBlur('year')}
+            value={values.year}
+          />
+
+          <Text>isPublic</Text>
+            <BouncyCheckbox
+              onPress={() => toggleIsPublic(values)}>
```
Comment (CodeGeneration-2020) [inline mobile/components/CreateTodo.tsx:67]: inline function

File: mobile/components/CreateTodo.tsx [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+            </BouncyCheckbox>
+
+          <Text>isComplete</Text>
+            <BouncyCheckbox 
+              onPress={() => toggleIsCompleted(values)}>
```
Comment (CodeGeneration-2020) [inline mobile/components/CreateTodo.tsx:72]: inline function

File: mobile/components/EditTodo.tsx [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  const toggleIsCompleted = (values: any) => {
+    values.isCompleted = !values.isCompleted
+  }
+
+  return (
+    <Formik
+      initialValues={initialValues}
+      onSubmit={values => {
```
Comment (CodeGeneration-2020) [inline mobile/components/EditTodo.tsx:33]: inline function

File: mobile/components/EditTodo.tsx [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+            style={styles.input}
+            onChangeText={handleChange('year')}
+            onBlur={handleBlur('year')}
+            value={values.year}
+          />
+
+          <Text>isPublic</Text>
+            <BouncyCheckbox onPress={() => toggleIsPublic(values)}></BouncyCheckbox>
+          <Text>isComplete</Text>
+            <BouncyCheckbox onPress={() => toggleIsCompleted(values)}></BouncyCheckbox>
```
Comment (CodeGeneration-2020) [inline mobile/components/EditTodo.tsx:70]: please avoid inline functions. everything should be named

File: mobile/components/EditTodo.tsx [hunk 28]
```diff
@@ -0,0 +1,93 @@
...
+          <Text>Year</Text>
+          <TextInput
+            style={styles.input}
+            onChangeText={handleChange('year')}
+            onBlur={handleBlur('year')}
+            value={values.year}
+          />
+
+          <Text>isPublic</Text>
+            <BouncyCheckbox onPress={() => toggleIsPublic(values)}></BouncyCheckbox>
```
Comment (CodeGeneration-2020) [inline mobile/components/EditTodo.tsx:68]: please avoid inline functions. everything should be named

File: mobile/components/Registration.tsx [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+    email: '',
+    password: '',
+    passwordVerification: '',
+  }
+
+  return (
+    <SafeAreaView style={styles.container}>
+      <Formik
+        initialValues={initialValues}
+        onSubmit={values => {}}
```
Comment (CodeGeneration-2020) [inline mobile/components/Registration.tsx:24]: ?? what is this?

File: mobile/components/SignIn.tsx [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+
+  const handleRegisterButton = () => {
+    navigation.push('Registration')
+  }
+
+  return (
+    <SafeAreaView style={styles.container}>
+      <Formik
+        initialValues={initialValues}
+        onSubmit={values => {}}
```
Comment (CodeGeneration-2020) [inline mobile/components/SignIn.tsx:31]: ?

**Review by CodeGeneration-2020 — CHANGES_REQUESTED** [review]
- Inline functions still exist
- Some props are handled incorrectly
- Always leave comments for potentially problematic parts of the code

**Pull request closed without merging**